### ISSUE TITLE
[WAGON-545] Connection reset while downloading artifacts in cloud environment (Azure)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-wagon#605]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** opened **[WAGON-545](https://issues.apache.org/jira/browse/WAGON-545?redirect=false)** and commented

We are building a huge multi-module project in the azure cloud and experience random connection resets during our build due to probably the same issue as in WAGON-486. Since it is stated by Microsoft that they terminate connections that are idle for longer than four minutes (see solution comment [here](https://developercommunity.visualstudio.com/content/problem/357696/maven-project-build-failing-with-connection-reset.html), I set -Dmaven.wagon.httpconnectionManager.ttlSeconds=120. Also, disabling the http pool does not solve the problem. Attached are logs from out main build with logging turned on for the wagon components, as mentioned in the comment section of WAGON-486.

---

**Affects:** 3.2.0

**Attachments:**
- [log_10_310.zip](https://issues.apache.org/jira/secure/attachment/12955961/log_10_310.zip) (_3.27 MB_)
- [log_10_311.zip](https://issues.apache.org/jira/secure/attachment/12955990/log_10_311.zip) (_3.49 MB_)
- [log_10_325.zip](https://issues.apache.org/jira/secure/attachment/12956303/log_10_325.zip) (_3.54 MB_)
- [successful-build-ttl5.zip](https://issues.apache.org/jira/secure/attachment/12956690/successful-build-ttl5.zip) (_5.10 MB_)

**Issue Links:**
- [WAGON-603](https://issues.apache.org/jira/browse/WAGON-603) Connection reset error from Azure pipelines
 (_**"is duplicated by"**_)
- [WAGON-546](https://issues.apache.org/jira/browse/WAGON-546) HttpWagon TTL not set
 (_**"depends upon"**_)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

There are several issues here. You are using Maven 3.5.4, this won't do. Please rerun with Maven 3.6.0. Moreover, the second run does only contain INFO output. This isn't sufficient.

Waiting for a new debug log file.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

I can try and see how I can convince the maven azure task to use v3.6.0, but there is only one run. And this one is configured with the slf4j options.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

I think we have a misunderstanding regarding the maven version, as I explicitly instruct maven to use wagon 3.3.1 via extensions. That's your whole point about maven 3.6.0, right?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Since I cannot see that Wagon 3.3.1 is used, I must insist on Maven 3.6.0. Using Wagon HTTP via extensions won't suffice because we need the updated API too. Moreover, one needs to use the shaded JAR. Please use 3.6.0.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

Ok, changed to maven 3.6.0, got rid of the extensions and have finally a good log. This is the commandline that maven was started with (is also visible inside the log btw):

```
mvn -B -U -s settings.xml -Dmaven.wagon.httpconnectionManager.ttlSeconds=120 -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient=DEBUG -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient.wire=ERROR -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient.impl.conn=DEBUG -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient.impl.client=DEBUG -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient.client=DEBUG install
```

Possible interesting information for you, Michael: the reason why the debug logs stopped once the reactor started in the previous run is due to the fact that I used the extension mechanism to pull in wagon 3.3.1. So, this tampers somehow with the logging information. Not sure if this useful information for you.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

> Possible interesting information for you, Michael: the reason why the debug logs stopped once the reactor started in the previous run is due to the fact that I used the extension mechanism to pull in wagon 3.3.1. So, this tampers somehow with the logging information. Not sure if this useful information for you.

This pretty much explains why the log was shut off. Maven uses `wagon-htttp-shaded`. HttpClient has different package names, that's why the logging went away.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Looking into the logs. Looks fishy because the reset happens on Maven Central. Can you add the thread name also. Log statements are interleaving..

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

How do I add the thread name?

(edit): found it. Running now, will upload when ready.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

New one with thread names enabled added (log_10_311.zip).

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks, will look into it a tad later.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

What I also don't get is why the retry handler does not kick in: https://github.com/apache/maven-wagon/commit/f920bc1a4d0f407453d0c45b5cf39316d9eb19d3. OK, it is because `SocketException` is of `ConnectException`.

[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch), can you reconfigure the retryhandler via properties to make `ConnectException` [retryable|#L96-L98]?

**Comment by jira-importer** [comment]
**[Oleg Kalnichevski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olegk)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I will have a closer look in the coming days.

Oleg

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Oleg Kalnichevski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olegk), thank you very much. It does not add up for me at the moment.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I am not sure, if I am able to setup wireshark in the azure cloud. Can you point me to some guidelines or documentation where I can get information how to do that?

For the retry, I will setup a new build later and keep you informed.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You must install it first on the node/server running Maven.. Start wireshark, listen on the interface, run maven, save tcpdump file. Try on your local machine first.

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

The thing is, in Azure I have no direct access to the machine. I can try do to certain stuff via scripting, but that's not very easy. Also, I am concerned about the size of the wireshark log. The process typically runs for 30-60mins before it fails.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think what you can do is do run `tcpdump` ony and filter on IP addresses, unles Oleg can find something w/o a dump this is going to be very hard...

**Comment by jira-importer** [comment]
**[Oleg Kalnichevski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olegk)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o): 

> What baffles me is that between the last get and the closure there is a 20 second gap and no response
> That is probably OK. The client was trying to send packets to the opposite endpoint and was blocked in the write operation until the connection get `RST`-ed. Please note that in the blocking i/o the socket timeout apples to read operations only.

> The connection should have been invalidated and not reused.
> Yes, it clearly should have been discarded and never re-used. This looks troubling.

What version of HttpClient is being used by the Maven Wagon in this particular case? Is there any chance it could be upgraded to 4.5.7?

Oleg

**Comment by jira-importer** [comment]
**[Oleg Kalnichevski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olegk)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)  [Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch): That is reasonably new. No need to upgrade. I will see if I can glean anything from the original log.

Oleg

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I am just trying to setup the retry handler. Looking at the configurations, I am not sure how I make ConnectionException retryable. Shouldn't it be by default? I can setup **non**-retryable classes. But how should I tell wagon something is retryable?

**Comment by jira-importer** [comment]
**[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch)** commented

I added additional logs. What I added to the startup are those configs:

```
-Dmaven.wagon.http.retryHandler.class=default -Dmaven.wagon.http.retryHandler.requestSentEnabled=true -Dmaven.wagon.http.retryHandler.count=5 
```

Still working on the tcpdump stuff.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

That's the whole point. No connection has ever expired. That's weird. Is the linked code correct to configure TTL for connections? I will double-check the TTL to work or not.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Christian Domsch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cdomsch), [here](http://home.apache.org/~michaelo/WAGON-546/apache-maven-3.6.1-SNAPSHOT-bin.zip) is a patched version. Please try that one. Don't change TTL, 5 min should be just fine as well as default retry config.

**Comment by jira-importer** [comment]
**[Oleg Kalnichevski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olegk)** commented

> why do we have connection resets on Maven Central? It is either the outboud connection with the NAT on Azure or the inbound reverse proxy of Maven Central has a 5 min cap without a proper keep-alive header
> [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) Connections resets can happen for all sorts of reasons, connections dropped due to inactivity being just one. Ideally the HTTP transport code should be resilient enough to recover from such I/O failures. In case of HttpWagon that uses HTTP GET to download artifacts re-executing requests failed due to `SocketException` should be the simplest and most reasonable thing to do.

Oleg

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed implicitly by Fixed with [dfd22586b6b934aa5a652ca32d57ed26067432fd|https://gitbox.apache.org/repos/asf?p=maven-wagon.git;a=commit;h=dfd22586b6b934aa5a652ca32d57ed26067432fd. Azure users shall set the TTL to 240 seconds or less.
